src/System.Memory/src/System/Buffers/BuffersExtensions.cs

```diff
@@ -38,10 +58,33 @@ public static void CopyTo<T>(this ReadOnlySequence<T> sequence, Span<T> destinat
             if (sequence.Length > destination.Length)
                 ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.destination);
 
-            foreach (var segment in sequence)
+            if (sequence.IsSingleSegment)
+            {
+                sequence.First.Span.CopyTo(destination);
+            }
+            else
+            {
+                CopyToMultiSegement(sequence, destination);
+            }
+        }
+
+        private static void CopyToMultiSegement<T>(this ReadOnlySequence<T> sequence, Span<T> destination)
+        {
+            var enumerator = sequence.GetEnumerator();
+            if (enumerator.MoveNext())
             {
-                segment.Span.CopyTo(destination);
-                destination = destination.Slice(segment.Length);
+                while (true)
+                {
+                    var segment = enumerator.Current.Span;
+                    segment.CopyTo(destination);
+                    if (!enumerator.MoveNext())
+                    {
+                        // Don't slice again if no more to copy
+                        break;
```

Comment:
Should this return instead of break? You'll exist out of the while loop but you'll call MoveNext again unnecessarily.

Reply:
Shouldn't do? Is `if` at top. Removed the enumerator though.